File: UnityBridge/com.opuscope.bridge/Runtime/BridgeWorkflowProtocol.cs
using System;
using Newtonsoft.Json;

namespace Opuscope.Bridge
{
    public class WorkflowCompletion
    {
        public static string Path => "/workflow/completed";
    
        [JsonProperty("identifier")]
        public string Identifier { get; set; }
    
        [JsonProperty("result")]
        public string Result { get; set; }

        public static WorkflowCompletion Make(string identifier, string result) => new ()
        {
            Identifier = identifier,
            Result = result
        };
    }

    public class WorkflowFailure
    {
        public static string Path => "/workflow/failed";
    
        [JsonProperty("identifier")]
        public string Identifier { get; set; }
    
        [JsonProperty("type")]
        public string Type { get; set; }
    
        [JsonProperty("message")]
        public string Message { get; set; }
        
        public static WorkflowFailure Make(string identifier, Exception e)
        {
            return new WorkflowFailure();
        }

        public Exception ToException()
        {
            return new Exception();
        }
    }

    public class WorkflowRequest
    {
        public static string Path => "/workflow/request";
    
        [JsonProperty("identifier")]
        public string Identifier { get; set; }
    
        [JsonProperty("procedure")]
        public string Procedure { get; set; }
    
        [JsonProperty("payload")]
        public string Payload { get; set; }
    }

    public class WorkflowCancellation
    {
        public static string Path => "/workflow/cancel";
    
        [JsonProperty("identifier")]
        public string Identifier { get; set; }
    }
}

File: SwiftBridge/Sources/SwiftBridge/BridgeWorkflowProtocol.swift
//
//  BridgeWorkflow.swift
//  
//
//  Created by Jonathan Thorpe on 05/06/2023.
//

import Foundation

public enum WorkflowError : Error {
    case noBridge
    case invalidProcedure(String)
    case runtime(type: String, message: String)
    case unknown
}

struct WorkflowCompletion : Codable {
    static let path = "/workflow/completed"
    let identifier : String
    let result : String
}

struct WorkflowFailure : Codable {
    static let path = "/workflow/failed"
    static let cancellationType = "CancelledWorkflow"
    static let invalidType = "InvalidWorkflow"
    let identifier : String
    let type : String
    let message : String
    
    func toError() -> Error {
        switch type {
        case WorkflowFailure.invalidType:
            return WorkflowError.invalidProcedure(message)
        case WorkflowFailure.cancellationType:
            return CancellationError()
        default:
            return WorkflowError.runtime(type: type, message: message)
        }
    }
    
    static func from(identifier: String, error : Error) -> WorkflowFailure {
        switch error {
        case is CancellationError:
            return WorkflowFailure(identifier: identifier, type: WorkflowFailure.cancellationType, message: "")
        case let WorkflowError.invalidProcedure(procedure):
            return WorkflowFailure(identifier: identifier, type: WorkflowFailure.invalidType, message: procedure)
        case let WorkflowError.runtime(type, message):
            return WorkflowFailure(identifier: identifier, type: type, message: message)
        default:
            return WorkflowFailure(identifier: identifier, type: "", message: "")
        }
    }
}

struct WorkflowRequest : Codable {
    static let path = "/workflow/request"
    let identifier : String
    let procedure : String
    let payload : String
}

struct WorkflowCancellation : Codable  {
    static let path = "/workflow/cancel"
    let identifier : String
}
